### ISSUE TITLE
Conflicting setter definitions for property \“currency\”: javax.money.MonetaryAmountFactory

### ISSUE THREAD

**Issue by vijayaratha** [FasterXML/jackson-databind#1251]
I use com.fasterxml.jackson and  io.swagger libraries. In my REST endpoint I use org.javamoney.moneta.Money type for a GET query. When deploying the war i get following exception ;

`Caused by: java.lang.IllegalArgumentException: Conflicting setter definitions for property "currency": javax.money.MonetaryAmountFactory#setCurrency(1 params) vs javax.money.MonetaryAmountFactory#setCurrency(1 params)
        at com.fasterxml.jackson.databind.introspect.POJOPropertyBuilder.getSetter(POJOPropertyBuilder.java:293)
        at io.swagger.jackson.ModelResolver.resolve(ModelResolver.java:246)
        at io.swagger.jackson.ModelResolver.resolve(ModelResolver.java:127)
        at io.swagger.converter.ModelConverterContextImpl.resolve(ModelConverterContextImpl.java:99)
        at io.swagger.jackson.ModelResolver.resolveProperty(ModelResolver.java:106)
        a`

I have followed this reference (http://stairs-jumper.livejournal.com/33244.html)  and wrote following code snippet[2]; and registered it at @ApplicationPath. But still getting same issue.

```java
    import javax.money.CurrencyUnit;
    import javax.money.Monetary;
    import javax.money.MonetaryAmountFactory;
    import javax.ws.rs.ext.Provider;
    import javax.xml.bind.annotation.XmlTransient;

    import com.fasterxml.jackson.annotation.JsonIgnore;
    import com.fasterxml.jackson.databind.ObjectMapper;
    import com.fasterxml.jackson.jaxrs.json.JacksonJsonProvider;

    @Provider
    public class JsonMoneyProvider extends JacksonJsonProvider {
        public JsonMoneyProvider() {
            ObjectMapper mapper = new ObjectMapper();
            mapper.addMixIn(MonetaryAmountFactory.class, MixIn.class);
            setMapper(mapper);
        }

        public static interface MixIn {
            @JsonIgnore
            @XmlTransient
            MonetaryAmountFactory setCurrency(CurrencyUnit currency);

            @JsonIgnore
            @XmlTransient
            default MonetaryAmountFactory setCurrency(String currencyCode) {
                return setCurrency(Monetary.getCurrency(currencyCode));
            }
        }
    }
```

Registering;

```java
@ApplicationPath("/rest")
public class RestApplication extends Application {
@Override
    public Set<Class<?>> getClasses() {
        HashSet<Class<?>> set = new HashSet<Class<?>>();
        set.add(com.test.JsonMoneyProvider.class);
```

SO query

http://stackoverflow.com/questions/37452689/how-to-overcome-conflicting-setter-definitions-for-property

**Comment by cowtowncoder** [comment]
First: have you tried latest Jackson version (2.7.4)?

**Comment by vijayaratha** [comment]
Yes..Core,databind,jsr librariess are 2.7.4 version.
And I try to avoid the conflicting methods using MixIn class;


```
import javax.money.CurrencyUnit;
import javax.money.MonetaryAmountFactory;
import javax.ws.rs.ext.Provider;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.jaxrs.json.JacksonJsonProvider;

@Provider
public class JsonMoneyProvider extends JacksonJsonProvider {
	public JsonMoneyProvider() {
		ObjectMapper mapper = new ObjectMapper();
		mapper.addMixIn(MonetaryAmountFactory.class, MixIn.class);
		setMapper(mapper);
	}

	public static interface MixIn {
		@JsonIgnore
		MonetaryAmountFactory setCurrency(CurrencyUnit currency);

		@JsonIgnore
		MonetaryAmountFactory setCurrency(String currencyCode);
	}
}

```

The error i get is;

`09:35:31,046 ERROR [org.jboss.msc.service.fail] (ServerService Thread Pool -- 163) MSC000001: Failed to start service jboss.undertow.deployment.default-server.default-host./test2dbwar: org.jboss.msc.service.StartException in service jboss.undertow.deployment.default-server.default-host./test2dbwar: java.lang.IllegalArgumentException: Conflicting setter definitions for property "currency": javax.money.MonetaryAmountFactory#setCurrency(1 params) vs javax.money.MonetaryAmountFactory#setCurrency(1 params)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentService$1.run(UndertowDeploymentService.java:85)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
	at org.jboss.threads.JBossThread.run(JBossThread.java:320)
Caused by: java.lang.IllegalArgumentException: Conflicting setter definitions for property "currency": javax.money.MonetaryAmountFactory#setCurrency(1 params) vs javax.money.MonetaryAmountFactory#setCurrency(1 params)
	at com.fasterxml.jackson.databind.introspect.POJOPropertyBuilder.getSetter(POJOPropertyBuilder.java:293)
	at io.swagger.jackson.ModelResolver.resolve(ModelResolver.java:246)
	at io.swagger.jackson.ModelResolver.resolve(ModelResolver.java:127)
	at io.swagger.converter.ModelConverterContextImpl.resolve(ModelConverterContextImpl.java:99)
	at io.swagger.jackson.ModelResolver.resolveProperty(ModelResolver.java:106)
	at io.swagger.jackson.ModelResolver.resolveProperty(ModelResolver.java:71)
	at io.swagger.converter.ModelConverterContextImpl.resolveProperty(ModelConverterContextImpl.java:79)
	at io.swagger.jackson.ModelResolver.resolve(ModelResolver.java:342)
	at io.swagger.jackson.ModelResolver.resolve(ModelResolver.java:127)
	at io.swagger.converter.ModelConverterContextImpl.resolve(ModelConverterContextImpl.java:99)
	at io.swagger.jackson.ModelResolver.resolveProperty(ModelResolver.java:106)
	at io.swagger.jackson.ModelResolver.resolveProperty(ModelResolver.java:71)
	at io.swagger.converter.ModelConverterContextImpl.resolveProperty(ModelConverterContextImpl.java:79)
	at io.swagger.jackson.ModelResolver.resolve(ModelResolver.java:342)
	at io.swagger.jackson.ModelResolver.resolve(ModelResolver.java:127)
	at io.swagger.converter.ModelConverterContextImpl.resolve(ModelConverterContextImpl.java:99)
	at io.swagger.jackson.ModelResolver.resolveProperty(ModelResolver.java:106)
	at io.swagger.jackson.ModelResolver.resolveProperty(ModelResolver.java:71)
	at io.swagger.converter.ModelConverterContextImpl.resolveProperty(ModelConverterContextImpl.java:79)
	at io.swagger.jackson.ModelResolver.resolve(ModelResolver.java:342)
	at io.swagger.jackson.ModelResolver.resolve(ModelResolver.java:127)
	at io.swagger.converter.ModelConverterContextImpl.resolve(ModelConverterContextImpl.java:99)
	at io.swagger.jackson.ModelResolver.resolveProperty(ModelResolver.java:106)
	at io.swagger.jackson.ModelResolver.resolveProperty(ModelResolver.java:71)
	at io.swagger.converter.ModelConverterContextImpl.resolveProperty(ModelConverterContextImpl.java:79)
	at io.swagger.converter.ModelConverters.readAsProperty(ModelConverters.java:58)
	at io.swagger.util.ParameterProcessor.applyAnnotations(ParameterProcessor.java:175)
	at io.swagger.jaxrs.Reader.getParameters(Reader.java:914)
`

**Comment by cowtowncoder** [comment]
Ok that seems like a problem. But looking at stack trace, this seems to be via Swagger set up -- is it possible it might use different `ObjectMapper` than one you are using?

Another thing that looks odd is the type itself -- why is factory type being serialized? Is that intentional? Usually only value types are: and in this case, there doesn't even seem to be `getCurrency()` method?

**Comment by vijayaratha** [comment]
I see setCurreny() method gives issue ?(based on the error stack) and when i check  " javax.money" API it has been defined like that. So i annotated like that. It is wrong? If so, what would be the correct way to ignore those setCurrency()?

**Comment by cowtowncoder** [comment]
@vijayaratha annotation on mix-in seems fine in general (there are some aspects that should be change, like only annotation one setter, not both, but the general idea is sound).

But as I said I don't think the issue is definition of mix-in as much as mix-in not being used; or that the type referred should not even be used: factories are usually not serialized at all.

As to JAX-RS code, Swagger; I don't know. This is outside of Jackson control and debugging those registrations is something you'd need to discuss with others.

**Comment by cowtowncoder** [comment]
At this point I would need a stand-alone test case to show how this occurs outside of JAX-RS/Swagger; otherwise it seems to be problem with configuring swagger to use properly configure `ObjectMapper`.

May be reopened with a stand-alone test case showing the issue.

**Comment by mykolapolonskyi** [comment]
Maybe will be useful even with this huge delay:
If you`re using `swagger-maven-plugin` validate version. I`ve fount that in version 3.1.4 is very old version of FaxterXML(1.9.X branch) 
Fix for it already placed in the 3.1.5-SNAPSOT of this plugin
